README
Sunflow Global Illumination Rendering System 
v0.07.5

Contact: Christopher Kulla
         fpsunflower@users.sourceforge.net
         http://sunflow.sourceforge.net/

----------------------------------------------------------------

Sunflow is a rendering system for photo-realistic image synthesis. It is written in Java and built around a flexible ray tracing core and an extensible object-oriented design.

Please consult the LICENSE file for license information.

----------------------------------------------------------------

Quickstart:

The fastest way to get started rendering images if you are not familiar with Java development is to get the binary release from the website. You will also need the latest Java SE JDK (to get the server VM) from http://java.sun.com/.

Launch sunflow from the command prompt like this:

>> java -server -Xmx1024M -jar sunflow.jar

IMPORTANT PERFORMANCE TIPS:
* Change the -Xmx command line option to match the amount of RAM in your system (example uses 1GB)
* Make sure the server VM is being used. Unfortunately, on 32 bit Windows, Sun only ships the server VM with the JDK, and doesn't install it in your path by default. Check your PATH environment variable if you have trouble.

Add -h to the end of command line for a list of command line options.

You can find some simple demo scenes on the website's data distribution. A simple blender script is also available from the website.

DISCLAIMER:
Keep in mind that this is still an early version. At the moment you will need to dig around the scene files in order to get the most out of the software. If you have any questions, feel free to e-mail me at the adress at the top.

----------------------------------------------------------------

Build instructions:

Download the latest JDK (6.0 at the time of this writing) if you don't have it already. Please note that the source code makes use of some new features like generics. Keep this in mind if you are trying to compile/run the code under a different JVM.

Create a main directory to hold the code (for example, "sunflow") and unzip the contents of the source package into it, preserving sub-directories. Create the "classes" subdirectory if your unzip program did not. You may now compile the software from the main directory by running:

>> javac -classpath janino.jar -sourcepath src -d classes -g:none -O src/*.java

once the compiling is complete, run the code with:

>> java -cp classes;janino.jar -server -Xmx1024M SunflowGUI

The tips above apply here as well (-Xmx and -server command line options).

----------------------------------------------------------------

Scene file format:

The SunflowGUI program accepts input in the .sc file format. As this is only a temporary file format, the best documentation for it is SCParser.java. You may also get a feel for what is supported by examining the example scene files provided in the data distribution.

----------------------------------------------------------------

Rendering options:

Here is a quick explanation of the basic rendering options.

Anti-aliasing:

The most important controls are aa.min and aa.max. These control the under/over-sampling of the image. The image will be first sampled at the rate prescribed by aa.min. Then, based on color and surface normal differences, the image will be refined up to the rate prescribed by aa.max.

A value of 0 corresponds to 1 sample per pixel.
A value of -1 corresponds to 1 sample every 2 pixels (1 per 2x2 block)
A value of -2 corresponds to 1 sample every 4 pixels (1 per 4x4 block)
A value of -3 corresponds to 1 sample every 8 pixels (1 per 8x8 block)
...
A value of 1 corresponds to 4 samples per pixel (2x2 subpixel grid)
A value of 2 corresponds to 16 samples per pixel (4x4 subpixel grid)
A value of 3 corresponds to 64 samples per pixel (8x8 subpixel grid)
...

Examples:
    - quick undersampled preview:          -2 0
    - preview with some edge refinement:    0 1
    - final rendering:                      1 2

Filtering:

Image quality is also affected by filtering. If you use oversampling (positive min or max AA), you will want to turn this on. Here are the names of the built-in filters:
  * box
  * triangle
  * gaussian
  * catmull-rom
  * mitchell
  * lanczos
  * blackman-harris
  * sinc

Box and triangle are best for previews as they are small and fast. The other filters are recommended for final image rendering.

Bucket rendering:

Sunflow proceses the image to be rendered in small squares called buckets. The size of these buckets can be controlled by a pixel width. Each rendering thread will be assigned to a single bucket. You may not get the bucket size you expect if you try to make them really small or really big, as there are some hard-coded limits to prevent excessive memory usage or excessive overhead.

The bucket ordering simply affects the order in which the buckets appear. They shouldn't have too much of an effect on overall rendering speed.

Soft shadows:

Area lights automatically create soft shadows in your scene. Once you have the proper meshlight exported, you can control the quality of shadows by changing the samples value. Please note this value is a number of rays PER triangle. Complex meshes made up of many triangles can become quite costly. The best is to restrict yourself to simple quads (2 triangles).

Global illumination:

Sunflow includes several methods to achieve global illumination. The easiest to setup is path tracing.

Path tracing is controlled by a sampling value and a number of bounces. A good starting point is 16 samples and 1 bounce. Increase the samples as needed to make the image less noisy. Note that this technique is not generally usable for complex lighting setups.

Irradiance caching or IGI. Tutorials comming soon.

Caustics:

Caustics are produced by light shining through refractive objects or being bounced by highly reflective materials. The only caustic algorithm currently implemented is via photon mapping. You must first pick a number of photons to shoot into the scene. Each light source will shoot this many photons, so be carefull if you have many lights in the scene. Pay attention to the messages in the console to see exactly what is going on.

Once you have a number of photons to emit, you must pick a way to store them. Only a kd engine is currently available for caustics. You can then set a value for the number of photons to gather at each shading point (start with ~50 to ~100) as well as a maximum search radius. These settings are highly scene dependent so experiment with them until you get satisfactory results.

----------------------------------------------------------------

Third party libraries:

Sunflow makes use of the following libraries, distributed according to the following terms:

Janino - An embedded Java[TM] compiler

Copyright (c) 2006, Arno Unkrig
All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

   1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
   2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
   3. The name of the author may not be used to endorse or promote products derived from this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


CHANGELOG
v0.07.5
* Added procedural scenes
* Added box type
v0.07.4
* Updated janino version to 2.7.8
* Allow changing shading cache parameters for a quality/speed trade-off
* Fixed missing TriangleMeshLight shader in PluginRegistry
* Fixed self-intersection bias
* Fixed infinite loop in bih caused by small triangle
v0.07.3
* Fixed bump mapping to correctly treat black as the lowest point in the map
* Added command line override for samples paramter (affect bucket and multipas samplers)
* Added cubic bspline filter to multipass image sampler (via warp function)
* Fixed "usage of init" statistic report in BIH
* Added cubic bspline filter
* Fixed counting of secondary rays clipped by ray depth limit
* Added lattice sampling routines (experimental)
* Fixed CornellBox light source sampling
* Fixed all image readers/writer to use buffered I/O
* Added motion blur syntax for instances and objects in .sc files
* Added shutter parameter to camera to allow motion blur of scene objects even when the camera is static
* Camera and Instances are now controlled by moving matrices and may be motion blurred
* Added MovingMatrix4 class
* Removed eye/target/up syntax for describing camera transform (parameters only - not .sc syntax)
* Added total rays to statistics output
* Fixed filtering in y direction for BucketRenderer
* Fixes to QMC sigma table - no longer requires precomputation
* Implemented low sampling density for ImageBasedLight in fixed mode
* Added PerlinModifier
* Bug fix to writeVerbatimString() in binary scene writer
* Added sampler choice to image block in .sc parser
* Small bugfix to command line sampler choice
* Moved shading cache into new image sampler
* Added a new image sampler (MultipassRenderer - not multipass yet)
* Added statistics to count number of rays cast
* Removed "robust" intersection code for triangles and bezier patches
* Added sphereflake primitive
* Moved Nan/Inf check into LightServer, demoted to warning severity
* Fix to allow LDR images in IBL
* File browser in SunflowGUI remembers previous directory
* Modified Matrix4 class to only accept affine transformations
* Added cylinder primitive
* Removed getUniqueName method from SunflowAPI
* Fixed more potential locale issues with case conversion by forcing the ENGLISH locale to be used
* Renamed get/setCurrentFrame to currentFrame in SunflowAPI
* Renamed parse to include in SunflowAPI
* Added reset keyword to new scene file format
* Fixed access levels on certain classes in org.sunflow package
* Fixed spelling of absorption in glass shader - left backwards compatibility in SCParser
* Added reading/writing of Indigo's IGI file format
* Added GenericBitmap format for simple debugging tasks
* Added more robust file extension detection
* Refactored Bitmap loading/saving with plugin system
* Bug fix when reading interpolation types on non-english operating systems
* Added -translate command line parameter
* Added translate() method to convert any file format into the new ones
* New ascii/binary file formats (sca/scb)
* Added new colorspace options for color parameters (blackbody/spectrum)
* Changed the way colors are specified to SunflowAPI
* Upgraded to Janino 2.5.6
* Added camera shift option to pinhole and thinlens camera lenses
* Updated examples
* Moved scene parsers to plugin system
* Moved photon maps to plugin system
* Replaced object factories with PluginRegistry lookups
* Updated all parsers
* Added PluginRegistry and modified SunflowAPI to create objects through it instead of requiring users to create objects themselves
* Changed cornellbox primitive into the light block in the .sc parser
* Light objects can now specify their geometric represtentation at runtime
* Particle point positions can now be specified inline in the scene file
* Particle data files can now be scene relative (via include paths)
* Improved motion blur sampling
* Moved shaderOverride to an option

v0.07.2
* Filled in and corrected javadoc comments in org.sunflow.core package
* Upgraded to Janino 2.5.5
* Added experimental Ant build file
* Removed hard coded sRGB curve
* Fixes to bucket renderer anti-aliasing
* Camera motion blur
* Added bezier patch meshes to parser
* Added normal smoothing to FileMesh
* Implemented BVH2 case properly in BoundingIntervalHierarchy (needs benchmarking)
* Added binary STL format to FileMesh
* Upgraded to Janino 2.5.4
* Improvements to Geometry class to handle tesselation failures
* Created FileMesh class for on-demand loading of meshes (supports RA3/OBJ)
* Tweaks to Benchmark class for third parties
* Small bugfixes
* Added little_endian option for particle files

v0.07.1
* Added Modifier concept - used to implement bump/normal mapping at the moment
* Added bucket dumping debugging option to bucket render - saves out unfiltered buckets as they are rendered
* Fixed jitter option - added to SCParser image block
* Added basic absorbtion model to GlassShader
* Added ngon sampling to ThinLens for bokeh effects
* Renamed MeshLight to TriangleMeshLight
* Updated LightSource interface - lights must now do their own loop to take mutltiple samples
* Spherical triangle sampling for triangle lights
* New wireframe shader (-quick_wire command line option)
* Added calls to get triangle vertices
* Added camera space utility functions
* Upgraded to Janino 2.5.3
* Upgraded to Janino 2.5.2
* Added sun light to SunSkyLight - modified LightSource interface slightly
* Added support for "noinstance" keyword
* Unified instance handling in .sc parser
* Added -pathgi command line argument
* Added basic animation support
* Added SunSkyLight
* Added QuadMesh primitive type (using bilinear patches)
* Added real sphere lights
* Console display now outputs to stderr to allow stdout to be used for output redirection
* Added fisheye lens to .sc parser
* Added imgpipe display compatible with mental ray imgpipe protocol

src/org/sunflow/Benchmark.java
package org.sunflow;

import java.awt.image.BufferedImage;
import java.io.IOException;
import java.net.URL;

import javax.imageio.ImageIO;

import org.sunflow.core.Display;
import org.sunflow.core.display.FileDisplay;
import org.sunflow.core.display.FrameDisplay;
import org.sunflow.image.Color;
import org.sunflow.math.Matrix4;
import org.sunflow.math.Point3;
import org.sunflow.math.Vector3;
import org.sunflow.system.BenchmarkFramework;
import org.sunflow.system.BenchmarkTest;
import org.sunflow.system.UI;
import org.sunflow.system.UserInterface;
import org.sunflow.system.UI.Module;
import org.sunflow.system.UI.PrintLevel;

public class Benchmark implements BenchmarkTest, UserInterface, Display {
    private int resolution;
    private boolean showOutput;
    private boolean showBenchmarkOutput;
    private boolean saveOutput;
    private boolean showWindow;
    private int threads;
    private int[] referenceImage;
    private int[] validationImage;
    private int errorThreshold;

    public static void main(String[] args) {
        if (args.length == 0) {
            System.out.println("Benchmark options:");
            System.out.println("  -regen                        Regenerate reference images for a variety of sizes");
            System.out.println("  -bench [threads] [resolution] Run a single iteration of the benchmark using the specified thread count and image resolution");
            System.out.println("                                Default: threads=0 (auto-detect cpus), resolution=256");
            System.out.println("  -show                         Render the benchmark scene into a window without performing validation");
        } else if (args[0].equals("-regen")) {
            int[] sizes = { 32, 64, 96, 128, 256, 384, 512 };
            for (int s : sizes) {
                // run a single iteration to generate the reference image
                Benchmark b = new Benchmark(s, true, false, true);
                b.kernelMain();
            }
        } else if (args[0].equals("-bench")) {
            int threads = 0, resolution = 256;
            if (args.length > 1)
                threads = Integer.parseInt(args[1]);
            if (args.length > 2)
                resolution = Integer.parseInt(args[2]);
            Benchmark benchmark = new Benchmark(resolution, false, true, false, threads, false);
            benchmark.kernelBegin();
            benchmark.kernelMain();
            benchmark.kernelEnd();
        } else if (args[0].equals("-show")) {
            Benchmark benchmark = new Benchmark(512, true, true, false, 0, true);
            benchmark.kernelMain();
        }
    }

    public Benchmark() {
        this(384, false, true, false);
    }

    public Benchmark(int resolution, boolean showOutput, boolean showBenchmarkOutput, boolean saveOutput) {
        this(resolution, showOutput, showBenchmarkOutput, saveOutput, 0, false);
    }

    public Benchmark(int resolution, boolean showOutput, boolean showBenchmarkOutput, boolean saveOutput, int threads, boolean showWindow) {
        UI.set(this);
        this.resolution = resolution;
        this.showOutput = showOutput;
        this.showBenchmarkOutput = showBenchmarkOutput;
        this.saveOutput = saveOutput;
        this.showWindow = showWindow;
        this.threads = threads;
        errorThreshold = 6;
        // fetch reference image from resources (jar file or classpath)
        if (saveOutput)
            return;
        URL imageURL = Benchmark.class.getResource(String.format("/resources/golden_%04X.png", resolution));
        if (imageURL == null)
            UI.printError(Module.BENCH, "Unable to find reference frame!");
        UI.printInfo(Module.BENCH, "Loading reference image from: %s", imageURL);
        try {
            BufferedImage bi = ImageIO.read(imageURL);
            if (bi.getWidth() != resolution || bi.getHeight() != resolution)
                UI.printError(Module.BENCH, "Reference image has invalid resolution! Expected %dx%d found %dx%d", resolution, resolution, bi.getWidth(), bi.getHeight());
            referenceImage = new int[resolution * resolution];
            for (int y = 0, i = 0; y < resolution; y++)
                for (int x = 0; x < resolution; x++, i++)
                    referenceImage[i] = bi.getRGB(x, resolution - 1 - y); // flip
        } catch (IOException e) {
            UI.printError(Module.BENCH, "Unable to load reference frame!");
        }
    }

    public void execute() {
        // 10 iterations maximum - 10 minute time limit
        BenchmarkFramework framework = new BenchmarkFramework(10, 600);
        framework.execute(this);
    }

    private class BenchmarkScene extends SunflowAPI {
        public BenchmarkScene() {
            build();
            render(SunflowAPI.DEFAULT_OPTIONS, showWindow ? new FrameDisplay() : saveOutput ? new FileDisplay(String.format("resources/golden_%04X.png", resolution)) : Benchmark.this);
        }

        @Override
        public void build() {
            // settings
            parameter("threads", threads);
            // spawn regular priority threads
            parameter("threads.lowPriority", false);
            parameter("resolutionX", resolution);
            parameter("resolutionY", resolution);
            parameter("aa.min", -1);
            parameter("aa.max", 1);
            parameter("filter", "triangle");
            parameter("depths.diffuse", 2);
            parameter("depths.reflection", 2);
            parameter("depths.refraction", 2);
            parameter("bucket.order", "hilbert");
            parameter("bucket.size", 32);
            // gi options
            parameter("gi.engine", "igi");
            parameter("gi.igi.samples", 90);
            parameter("gi.igi.c", 0.000008f);
            options(SunflowAPI.DEFAULT_OPTIONS);
            buildCornellBox();
        }

        private void buildCornellBox() {
            // camera
            parameter("transform", Matrix4.lookAt(new Point3(0, 0, -600), new Point3(0, 0, 0), new Vector3(0, 1, 0)));
            parameter("fov", 45.0f);
            camera("main_camera", "pinhole");
            parameter("camera", "main_camera");
            options(SunflowAPI.DEFAULT_OPTIONS);
            // cornell box
            float minX = -200;
            float maxX = 200;
            float minY = -160;
            float maxY = minY + 400;
            float minZ = -250;
            float maxZ = 200;

            float[] verts = new float[] { minX, minY, minZ, maxX, minY, minZ,
                    maxX, minY, maxZ, minX, minY, maxZ, minX, maxY, minZ, maxX,
                    maxY, minZ, maxX, maxY, maxZ, minX, maxY, maxZ, };
            int[] indices = new int[] { 0, 1, 2, 2, 3, 0, 4, 5, 6, 6, 7, 4, 1,
                    2, 5, 5, 6, 2, 2, 3, 6, 6, 7, 3, 0, 3, 4, 4, 7, 3 };

            parameter("diffuse", null, 0.70f, 0.70f, 0.70f);
            shader("gray_shader", "diffuse");
            parameter("diffuse", null, 0.80f, 0.25f, 0.25f);
            shader("red_shader", "diffuse");
            parameter("diffuse", null, 0.25f, 0.25f, 0.80f);
            shader("blue_shader", "diffuse");

            // build walls
            parameter("triangles", indices);
            parameter("points", "point", "vertex", verts);
            parameter("faceshaders", new int[] { 0, 0, 0, 0, 1, 1, 0, 0, 2, 2 });
            geometry("walls", "triangle_mesh");

            // instance walls
            parameter("shaders", new String[] { "gray_shader", "red_shader",
                    "blue_shader" });
            instance("walls.instance", "walls");

            // create mesh light
            parameter("points", "point", "vertex", new float[] { -50, maxY - 1,
                    -50, 50, maxY - 1, -50, 50, maxY - 1, 50, -50, maxY - 1, 50 });
            parameter("triangles", new int[] { 0, 1, 2, 2, 3, 0 });
            parameter("radiance", null, 15, 15, 15);
            parameter("samples", 8);
            light("light", "triangle_mesh");

            // spheres
            parameter("eta", 1.6f);
            shader("Glass", "glass");
            sphere("glass_sphere", "Glass", -120, minY + 55, -150, 50);
            parameter("color", null, 0.70f, 0.70f, 0.70f);
            shader("Mirror", "mirror");
            sphere("mirror_sphere", "Mirror", 100, minY + 60, -50, 50);

            // scanned model
            geometry("teapot", "teapot");
            parameter("transform", Matrix4.translation(80, -50, 100).multiply(Matrix4.rotateX((float) -Math.PI / 6)).multiply(Matrix4.rotateY((float) Math.PI / 4)).multiply(Matrix4.rotateX((float) -Math.PI / 2).multiply(Matrix4.scale(1.2f))));
            parameter("shaders", "gray_shader");
            instance("teapot.instance1", "teapot");
            parameter("transform", Matrix4.translation(-80, -160, 50).multiply(Matrix4.rotateY((float) Math.PI / 4)).multiply(Matrix4.rotateX((float) -Math.PI / 2).multiply(Matrix4.scale(1.2f))));
            parameter("shaders", "gray_shader");
            instance("teapot.instance2", "teapot");
        }

        private void sphere(String name, String shaderName, float x, float y, float z, float radius) {
            geometry(name, "sphere");
            parameter("transform", Matrix4.translation(x, y, z).multiply(Matrix4.scale(radius)));
            parameter("shaders", shaderName);
            instance(name + ".instance", name);
        }
    }

    public void kernelBegin() {
        // allocate a fresh validation target
        validationImage = new int[resolution * resolution];
    }

    public void kernelMain() {
        // this builds and renders the scene
        new BenchmarkScene();
    }

    public void kernelEnd() {
        // make sure the rendered image was correct
        int diff = 0;
        if (referenceImage != null && validationImage.length == referenceImage.length) {
            for (int i = 0; i < validationImage.length; i++) {
                // count absolute RGB differences
                diff += Math.abs((validationImage[i] & 0xFF) - (referenceImage[i] & 0xFF));
                diff += Math.abs(((validationImage[i] >> 8) & 0xFF) - ((referenceImage[i] >> 8) & 0xFF));
                diff += Math.abs(((validationImage[i] >> 16) & 0xFF) - ((referenceImage[i] >> 16) & 0xFF));
            }
            if (diff > errorThreshold)
                UI.printError(Module.BENCH, "Image check failed! - #errors: %d", diff);
            else
                UI.printInfo(Module.BENCH, "Image check passed!");
        } else
            UI.printError(Module.BENCH, "Image check failed! - reference is not comparable");

    }

    public void print(Module m, PrintLevel level, String s) {
        if (showOutput || (showBenchmarkOutput && m == Module.BENCH))
            System.out.println(UI.formatOutput(m, level, s));
        if (level == PrintLevel.ERROR)
            throw new RuntimeException(s);
    }

    public void taskStart(String s, int min, int max) {
        // render progress display not needed
    }

    public void taskStop() {
        // render progress display not needed
    }

    public void taskUpdate(int current) {
        // render progress display not needed
    }

    public void imageBegin(int w, int h, int bucketSize) {
        // we can assume w == h == resolution
    }

    public void imageEnd() {
        // nothing needs to be done - image verification is done externally
    }

    public void imageFill(int x, int y, int w, int h, Color c, float alpha) {
        // this is not used
    }

    public void imagePrepare(int x, int y, int w, int h, int id) {
        // this is not needed
    }

    public void imageUpdate(int x, int y, int w, int h, Color[] data, float[] alpha) {
        // copy bucket data to validation image
        for (int j = 0, index = 0; j < h; j++, y++)
            for (int i = 0, offset = x + resolution * (resolution - 1 - y); i < w; i++, index++, offset++)
                validationImage[offset] = data[index].copy().toNonLinear().toRGB();
    }
}